### PR TITLE
8343827: RISC-V: set AlignVector as false if applicable to enable SLP

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review this patch?

SLP vectorization in C2 checks the data alignment optionally, if fast misaliged access is supported by CPU it will skip the alignment check in SLP, this is good for both C2 compile time and run time (as more loops can be potentially vectorized. And the performance of vector implementation is good enough, which was already verified by existing tests (array copy, crc32, ...)).
On the other hand, we have some existing tests which verify that generated IRs should contain some specific Nodes, with the intrdocution of Compact Object Header, some tests related to SLP was observed failure which is due to the data offset change which is due to header change in the JEP, by enabling C2 SLP when fast misaliged access is supported, we should be able to resolve these issues. Although riscv support misaligned access optionally, started from RVA22 it's mandatory.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343827](https://bugs.openjdk.org/browse/JDK-8343827): RISC-V: set AlignVector as false if applicable to enable SLP (**Enhancement** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - Committer) Review applies to [696972ec](https://git.openjdk.org/jdk/pull/21974/files/696972ec364ec71ed5327c417c8651e35afc2285)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21974/head:pull/21974` \
`$ git checkout pull/21974`

Update a local copy of the PR: \
`$ git checkout pull/21974` \
`$ git pull https://git.openjdk.org/jdk.git pull/21974/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21974`

View PR using the GUI difftool: \
`$ git pr show -t 21974`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21974.diff">https://git.openjdk.org/jdk/pull/21974.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21974#issuecomment-2464526629)
</details>

### CODE REVIEW

**Review by RealFYang — APPROVED** [review]
Looks fine. Thanks.

**Comment by VladimirKempik** [comment]
hello Hamlin, this could to be wrong.
for example: kendryte 230 with c908 core with rv64gcv 
it has fast misaligned access but only for scalars(rv64gc), not for vectors(rvv)
i dont know if it's just a corner case or a possible common practice in future.

**Comment by RealFYang** [comment]
> hello Hamlin, this could to be wrong. for example: kendryte 230 with c908 core with rv64gcv it has fast misaligned access but only for scalars(rv64gc), not for vectors(rvv) i dont know if it's just a corner case or a possible common practice in future.

Interesting. Could it be an issue for the kernel running on this platform? I don't understand why it would declare fast misaligned access when it's not the case for the vector part.

**Comment by VladimirKempik** [comment]
> > hello Hamlin, this could to be wrong. for example: kendryte 230 with c908 core with rv64gcv it has fast misaligned access but only for scalars(rv64gc), not for vectors(rvv) i dont know if it's just a corner case or a possible common practice in future.
> 
> Interesting. Could it be an issue for the kernel running on this platform? I don't understand why it would declare fast misaligned access when it's not the case for the vector part.

for this particular case, it has kernel 5.10 which has no hwprobe. so avoidUnaligned has to be disabled manually. 
but there are more boards based on c908 core which needs to be checked with this change. 
while the c908 isnt the thing we should care much about, right now i think boards with c908 are the main test mule to check  jdk rvv performance changes

**Comment by Hamlin-Li** [comment]
> hello Hamlin, this could to be wrong. for example: kendryte 230 with c908 core with rv64gcv it has fast misaligned access but only for scalars(rv64gc), not for vectors(rvv) i dont know if it's just a corner case or a possible common practice in future.

Thanks for bring up this test case.
I have the similar question in mind, if a hardware claims to support misaligned fast access, seems it should support both scalar and vector.

> for this particular case, it has kernel 5.10 which has no hwprobe. so avoidUnaligned has to be disabled manually.

Yes, the default behaviour is to use hwprobe to test the CPU features, so for k 230, the default behaviour should be fine, i.e. both `AlignVector` and `AvoidUnalignedAccesses` would be true, nothing changed.

> but there are more boards based on c908 core which needs to be checked with this change. while the c908 isnt the thing we should care much about, right now i think boards with c908 are the main test mule to check jdk rvv performance changes

I'm not quite sure if we should test on each of the potential boards, seems to me we should not. What we need do should be to test CPU features based on hwprobe if unaligned fast access is supported or not, and enable/disable JDK features accordingly. In this way, our tests should be easy and get most of performance gain for common cases, and if there is a cornor case (claim to support but does not actaully), then the user could still disable it manually.

I'll wait for other tests/comments ...

File: src/hotspot/cpu/riscv/vm_version_riscv.cpp [hunk 1]
```diff
@@ -166,11 +166,12 @@ void VM_Version::common_initialize() {
...
-      FLAG_SET_DEFAULT(AvoidUnalignedAccesses, true);
-    } else {
-      FLAG_SET_DEFAULT(AvoidUnalignedAccesses, false);
-    }
+    FLAG_SET_DEFAULT(AvoidUnalignedAccesses,
+      unaligned_access.value() != MISALIGNED_FAST);
+  }
+
+  if (FLAG_IS_DEFAULT(AlignVector)) {
+    AlignVector = AvoidUnalignedAccesses;
```
Comment (luhenry) [inline src/hotspot/cpu/riscv/vm_version_riscv.cpp:174]: Nit: Should that be `FLAG_SET_DEFAULT(AlignVector, AvoidUnalignedAccesses)`?

**Review by luhenry — APPROVED** [review]

**Comment by luhenry** [comment]
> > for this particular case, it has kernel 5.10 which has no hwprobe. so avoidUnaligned has to be disabled manually.
>
> Yes, the default behaviour is to use hwprobe to test the CPU features, so for k 230, the default behaviour should be fine, i.e. both AlignVector and AvoidUnalignedAccesses would be true, nothing changed.

Should we argue that hwprobe should have two flags, one for scalar and one for vector? We can talk to some hardware folks for that

**Comment by Hamlin-Li** [comment]
> Should we argue that hwprobe should have two flags, one for scalar and one for vector? We can talk to some hardware folks for that

Good question, let's discuss later.

**Comment by Hamlin-Li** [comment]
@RealFYang @luhenry just made a minor improvement, can you take another look?

**Review by RealFYang — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
Thanks for your reviewing!

/integrate
